scipy/stats/stats.py

```diff
@@ -5670,21 +5705,145 @@ def ttest_ind(a, b, axis=0, equal_var=True, nan_policy='propagate'):
     if a.size == 0 or b.size == 0:
         return _ttest_nans(a, b, axis, Ttest_indResult)
 
-    v1 = np.var(a, axis, ddof=1)
-    v2 = np.var(b, axis, ddof=1)
-    n1 = a.shape[axis]
-    n2 = b.shape[axis]
+    if permutations is not None:
+        res = _permutation_ttest(a, b,
+                                 axis=axis,
+                                 equal_var=equal_var,
+                                 permutations=permutations,
+                                 random_state=random_state)
 
-    if equal_var:
-        df, denom = _equal_var_ttest_denom(v1, n1, v2, n2)
     else:
-        df, denom = _unequal_var_ttest_denom(v1, n1, v2, n2)
-
-    res = _ttest_ind_from_stats(np.mean(a, axis), np.mean(b, axis), denom, df)
+        v1 = np.var(a, axis, ddof=1)
+        v2 = np.var(b, axis, ddof=1)
+        n1 = a.shape[axis]
+        n2 = b.shape[axis]
 
+        if equal_var:
+            df, denom = _equal_var_ttest_denom(v1, n1, v2, n2)
+        else:
+            df, denom = _unequal_var_ttest_denom(v1, n1, v2, n2)
+        res = _ttest_ind_from_stats(np.mean(a, axis),
+                                    np.mean(b, axis),
+                                    denom, df)
     return Ttest_indResult(*res)
 
 
+def _broadcast_concatenate(xs, axis):
+    """Concatenate arrays along an axis with broadcasting"""
+    # move the axis we're concatenating along to the end
+    xs = [np.swapaxes(x, axis, -1) for x in xs]
+    # determine final shape of all but the last dim
+    shape = np.broadcast(*[x[..., 0] for x in xs]).shape
+    # get final size of the dimension we're concatenating along
+    length = sum((x.shape[-1] for x in xs))
+    # create an empty array of the right shape
+    res = np.empty(shape + (length,))
+    # fill it up
+    index = 0
+    for x in xs:
+        res[..., index: index+x.shape[-1]] = x
+        index += x.shape[-1]
+    # move the axis we concatenated along to where it was
+    res = np.swapaxes(res, axis, -1)
+    return res
+
+
+def _data_permutations(data, n=-1, axis=-1, random_state=None):
```

Comment:
Previously the code permuted the data once every loop iteration. This vectorizes the permutation generation, and evaluation of the t-statistic is also vectorized. It takes more memory, but it was supposed to be faster, and I think it could be really speedy if we allow CuPy backend. 

In the end, though, performance is only comparable between the two, and the bottleneck is generating all the random permutations of indices. I'll take a closer look at the code in `_perm_test` used in `multiscale_graphcorr` to see if it is faster, but probably not for just one worker - it uses list comprehension to initialize one `RandomState` object per permutation.

Is there a function that can generate ND arrays like:
```
[[4, 1, 2, 3]
 [1, 4, 3, 2]]
]
where each row (in this case) is a random permutation of the same set of numbers? Alternatively, can we generate a random 1/0 array in which each row has the same sum? Either of these could replace:
```
indices = rankdata(random_state.random((m, n)), axis=0) - 1
```
and really speed this up.

Maybe we can request from NumPy? (And maybe we can extend the capabilities of `concatenate` to allow broadcasting?)

Reply:
Don’t try to outsmart NumPy with some hackery that can turn out to be less performant in the future. Rather, if there is a bottleneck, we can apply a tiny bit of Cython here, and get it up to C speed. We can even multithread the bottleneck with a tiny bit of Cython.

Reply:
Also note that we do not need to make it “fastest” in the first version. Just make it mathematically correct. We can optimize for speed later on.

Reply:
Would you consider revising those posts to ease the tone?
If the current code is offensive, would you recommend the list comprehension with np.permutation for now?

Reply:
So leave it as it is now, or revert to when this was a for loop?

Reply:
Sorry about the language. We Scandinavians have a way of speaking that is rather direct (which is why many find Linus T. offensive as well 😂).

Anyhow, what I mean is that a for loop on random_state.shuffle should be ok. If it is indeed a bottleneck of some concern, we can move this particular loop to Cython, and also make a specialized shuffle in Cython, later on. It will give maximum O(n) performance without the memory tradeof or any need for O(n log n) argsorting. I think this is a clear case where vectorizing in NumPy is not a good way to solve a performance issue. So just leave it as a simple for loop on random.shuffle and move it to Cython in a later PR. Just focus on getting it mathematically correct, so we have a working Python prototype.

Reply:
With vectorization we get a memory tradeoff and O(n log n) instead of O(n). Thus we make Cython to make this fast without these problems.

Reply:
And the way we retain O(n) with Cython is to use the C++ standard library function std::nth_element on the computed t-values. But we cannjust leave that out for now. Just make it work correctly, without any attempt to optimize :)